game/buildings.py

```diff
@@ -0,0 +1,113 @@
+from abc import ABC
+from dataclasses import dataclass
+from datetime import timedelta
+from typing import ClassVar
+
+
+@dataclass
+class Building(ABC):
+    MAX_LEVEL: ClassVar[int] = 1
+    level: int = 1
+    upgrade_time: timedelta = timedelta(minutes=1)
+
+    # Upgrade costs
+    wood_cost: int = 0
+    clay_cost: int = 0
+    iron_cost: int = 0
+
+    def get_upgrade_cost(self) -> dict:
+        return {
+            "wood": self.wood_cost * self.level,
+            "clay": self.clay_cost * self.level,
+            "iron": self.iron_cost * self.level,
+        }
+
+    def get_upgrade_time(self) -> timedelta:
+        return self.upgrade_time * self.level
+
+    def upgrade(self) -> None:
+        if self.level < self.MAX_LEVEL:
+            self.level += 1
+
+    def downgrade(self) -> None:
+        if self.level > 1:
+            self.level -= 1
+
+
+class ResourceBuilding(Building, ABC):
+    # Resources per second
+    wood_production: float = 0
+    clay_production: float = 0
+    iron_production: float = 0
+
+
+@dataclass
+class TownHall(Building):
+    MAX_LEVEL = 3
+    upgrade_time: timedelta = timedelta(minutes=10)
```

Comment:
We should consider variable upgrade time depending on building level. It's only additional feature and it's up to you whether you want to apply it or not